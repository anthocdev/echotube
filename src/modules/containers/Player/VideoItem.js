import React from "react";
import Card from "react-bootstrap/Card";

export default function VideoItem({ index, video, selectedVideo }) {
  return (
    <div
      onClick={() => {
        selectedVideo(index, video.VideoLink);
      }}
    >
      <Card className="videoCard">
        <Card.Img
          variant="top"
          src={`https://i.ytimg.com/vi/${video.VideoLink}/hqdefault.jpg`}
        />
        <Card.ImgOverlay>
          <Card.Title>{video.Name}</Card.Title>
        </Card.ImgOverlay>
      </Card>
    </div>
  );
}
