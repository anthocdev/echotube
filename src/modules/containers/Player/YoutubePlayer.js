import React from "react";
import Youtube from "react-youtube";
import { connect } from "react-redux";
import { setPlayingIndex } from "../../../actions/playback";

/* YouTube video player instance*/
class YoutubePlayer extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      index: 0
    };
  }

  componentDidMount() {}

  render() {
    //Defining options for YouTube player
    const opts = {
      height: 320,
      width: 540,
      playerVars: {
        autoplay: 1
      }
    };
    return (
      <div>
        <Youtube
          videoId={this.props.videoId} //Video ID
          opts={opts} //Options
          onReady={this.onReady} //When player is ready
          onEnd={this.onEnd} //When video ends
        />
      </div>
    );
  }

  onReady(event) {
    event.target.playVideo();
  }

  onEnd(event) {
    //To be implemented -> Play next video
  }
}

function mapStateToProps(state, props) {
  return {
    playbackItems: state.player.playbackItems,
    currentIndex: state.player.currentIndex
  };
}

function mapDispatchToProps(dispatch) {
  return {
    dispatch: function(task) {
      dispatch(task);
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(YoutubePlayer);
