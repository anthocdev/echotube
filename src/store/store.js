import { applyMiddleware, createStore, compose } from "redux";
import { createLogger } from "redux-logger";
import { default as thunk } from "redux-thunk";
import { createPromise } from "redux-promise-middleware";
import axios from "axios";

import reducer from "../reducers";
//Pulling JWT Token from local storage
const jwttoken = localStorage.getItem("JWT_TOKEN");
//Setting JWT Token as common header for axios
axios.defaults.headers.common["Authorization"] = jwttoken;
//Create store with middleware included
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose; //Adding redux dev tool compatibility
const middleware = applyMiddleware(createPromise(), thunk, createLogger()); //Adding thunk and console logging for testing as middleware

export default createStore(
  reducer,
  {
    //Initializing auth state if JWT token is available
    auth: {
      jwtToken: jwttoken,
      isAuth: jwttoken ? true : false
    }
  },
  composeEnhancers(middleware)
);
