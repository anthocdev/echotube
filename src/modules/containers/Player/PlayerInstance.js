import React from "react";
import VideoItem from "./VideoItem";
import CardColumns from "react-bootstrap/Card";
import * as PlayerActions from "../../../actions/PlayerActions";
import YoutubePlayer from "./YoutubePlayer";
import { connect } from "react-redux";
import { Button } from "@material-ui/core";

/* Player overlay with YouTube iframe and custom video listing column */

class Player extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  state = {
    currentlyPlaying: null,
    playIndex: null,
  };

  removeFromQueue = (index) => {
    this.props.RemovePlaybackItem(index);
  };

  selectVideo = (index, videoLink) => {
    this.props.SetActiveItem(index);

    this.setState({
      currentlyPlaying: videoLink,
      playIndex: index,
    });
  };

  UpdatePlayback = () => {
    if (this.props.PlaybackQueue[this.props.CurrentIndex] != undefined) {
      return this.props.PlaybackQueue[this.props.CurrentIndex].VideoLink;
    } else {
      return null;
    }
  };

  render() {
    const mapVideos = (
      <div>
        {this.props.player.PlaybackQueue.map((data, i) => (
          <CardColumns className="videoColumns">
            <VideoItem
              video={data}
              key={i}
              index={i}
              selectedVideo={this.selectVideo}
            />
            <Button onClick={() => this.removeFromQueue(i)}>Remove</Button>
          </CardColumns>
        ))}
      </div>
    ); //Generate video list by mapping data from playbackItems state

    const playerContainer = (
      <div className="playerMain">
        <div className="left-container">
          <YoutubePlayer videoId={this.UpdatePlayback()} />
        </div>
        <div className="right-container">
          <div className="playlistColumn">
            {this.props.player.PlaybackQueue.length > 0 ? (
              mapVideos
            ) : (
              <div>No videos available</div>
            )}
          </div>
        </div>
      </div>
    );

    return (
      <div>
        <div className="Main">{playerContainer}</div>
      </div>
    );
  }
}

function mapStateToProps(store, props) {
  return {
    player: store.player,
    PlaybackQueue: store.player.PlaybackQueue,
    CurrentIndex: store.player.CurrentIndex,
  };
}

export default connect(mapStateToProps, PlayerActions)(Player);
