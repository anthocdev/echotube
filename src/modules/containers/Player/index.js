import React from "react";
import VideoItem from "./VideoItem";
import CardColumns from "react-bootstrap/Card";
import YoutubePlayer from "./YoutubePlayer";
import { togglePlayer, setPlayingIndex } from "../../../actions/playback";
import { connect } from "react-redux";

/* Player overlay with YouTube iframe and custom video listing column */

class Player extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  state = {
    currentlyPlaying: null,
    playIndex: null
  };

  selectVideo = (index, videoLink) => {
    this.props.dispatch(setPlayingIndex(index));

    this.setState({
      currentlyPlaying: videoLink,
      playIndex: index
    });
  };

  togglePlayer() {
    this.setState({ showPlayer: !this.state.showPlayer });
  } //Toggling player visibility state

  render() {
    const playerVisible = this.props.player.isVisible;
    const mapVideos = (
      <div>
        {this.props.player.playbackItems.map((data, i) => (
          <CardColumns className="videoColumns">
            <VideoItem
              video={data}
              key={i}
              index={i}
              selectedVideo={this.selectVideo}
            />
          </CardColumns>
        ))}
      </div>
    ); //Generate video list by mapping data from playbackItems state

    const playerContainer = (
      <div className="playerMain">
        {" "}
        <div className="left-container">
          <YoutubePlayer
            videoId={this.state.currentlyPlaying}
            selectedVideo={this.selectVideo}
          />
        </div>
        <div className="right-container">
          <div className="playlistColumn">
            {this.props.player.playbackItems.length > 0 ? (
              mapVideos
            ) : (
              <div>No videos available</div>
            )}
          </div>
        </div>
      </div>
    );

    return (
      <div>
        <button
          className="PlayerButton"
          onClick={() => this.props.dispatch(togglePlayer(playerVisible))}
        >
          {/*Change playerVisible property*/}
          Toggle Player
        </button>
        <div className="Main">
          {playerVisible ? playerContainer : <div></div>}
          {/*Display player container based on playerVisible property*/}
        </div>
      </div>
    );
  }
}

function mapStateToProps(state, props) {
  return {
    player: state.player
  };
}

function mapDispatchToProps(dispatch) {
  return {
    dispatch: function(task) {
      dispatch(task);
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Player);
