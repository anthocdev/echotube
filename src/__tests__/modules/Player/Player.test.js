import Player from "../../../modules/containers/Player";
import React from "react";
import { Provider } from "react-redux";
import store from "../../../store/store";
import { create, act } from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

describe("Overall Player Module Testing", () => {
  let root;

  it("Renders Without Crashing", async () => {
    act(() => {
      root = create(
        <Provider store={store}>
          <Player />
        </Provider>
      );
    });
  });

  test("Renders Matches Snapshot", async () => {
    act(() => {
      root = create(
        <BrowserRouter>
          <Provider store={store}>
            <Player />
          </Provider>
        </BrowserRouter>
      );
    });
    expect(root).toMatchSnapshot();
    console.log(root.toJSON());
  });
});
