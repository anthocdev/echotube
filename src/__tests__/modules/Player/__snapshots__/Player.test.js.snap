// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overall Player Module Testing Renders Matches Snapshot 1`] = `
<div>
  <button
    aria-label="Open Player"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        variant="solid"
        viewBox="0 0 24 24"
      >
        <path
          d="M17 10.5V6H3v12h14v-4.5l4 4v-11l-4 4z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
</div>
`;
